<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>fxtools</title>
        <meta name="robots" content="noindex" />
        <!-- Custom HTML head -->
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">
        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">
        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->
    </head>
    <body>
        <!-- Provide site root to javascript -->
        <script type="text/javascript">
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script type="text/javascript">
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script type="text/javascript">
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script type="text/javascript">
            var html = document.querySelector('html');
            var sidebar = 'hidden';
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="introduction.html">Introduction</a></li><li class="chapter-item expanded affix "><a href="installation.html">Installation</a></li><li class="spacer"></li><li class="chapter-item expanded affix "><li class="part-title">Tools</li><li class="chapter-item expanded "><a href="tools/cat.html"><strong aria-hidden="true">1.</strong> cat</a></li><li class="chapter-item expanded "><a href="tools/clip.html"><strong aria-hidden="true">2.</strong> clip</a></li><li class="chapter-item expanded "><a href="tools/count.html"><strong aria-hidden="true">3.</strong> count</a></li><li class="chapter-item expanded "><a href="tools/extract_variable.html"><strong aria-hidden="true">4.</strong> extract-variable</a></li><li class="chapter-item expanded "><a href="tools/filter.html"><strong aria-hidden="true">5.</strong> filter</a></li><li class="chapter-item expanded "><a href="tools/fix.html"><strong aria-hidden="true">6.</strong> fix</a></li><li class="chapter-item expanded "><a href="tools/sort.html"><strong aria-hidden="true">7.</strong> sort</a></li><li class="chapter-item expanded "><a href="tools/reverse.html"><strong aria-hidden="true">8.</strong> reverse</a></li><li class="chapter-item expanded "><a href="tools/sample.html"><strong aria-hidden="true">9.</strong> sample</a></li><li class="chapter-item expanded "><a href="tools/sgrna_table.html"><strong aria-hidden="true">10.</strong> sgrna-table</a></li><li class="chapter-item expanded "><a href="tools/t2g.html"><strong aria-hidden="true">11.</strong> t2g</a></li><li class="chapter-item expanded "><a href="tools/take.html"><strong aria-hidden="true">12.</strong> take</a></li><li class="chapter-item expanded "><a href="tools/trim.html"><strong aria-hidden="true">13.</strong> trim</a></li><li class="chapter-item expanded "><a href="tools/unique.html"><strong aria-hidden="true">14.</strong> unique</a></li><li class="chapter-item expanded "><a href="tools/upper.html"><strong aria-hidden="true">15.</strong> upper</a></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light (default)</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">fxtools</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>
                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script type="text/javascript">
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="-fxtools-"><a class="header" href="#-fxtools-">[ <code>fxtools</code> ]</a></h1>
<h2 id="introduction"><a class="header" href="#introduction">Introduction</a></h2>
<p><code>fxtools</code> is a collection of <code>fastx</code> related utilities I've
needed to write for one reason or another to perform
various functions. 
The tool is a collection of sequence processing utilities
written in rust and focusing generally on sequence
transformations.</p>
<p><code>fastx</code> is a term that means <em>either</em> FASTA or FASTQ.
The utilities in this tool generally do not care which
format the sequences are in, and will handle them
appropriately.</p>
<p>All subcommands within the tool are documented here, but
should be usable and understandable from their help menus.</p>
<p><code>fxtools</code> has been written with the unix-philosophy in mind
and most functions can be used with standard CLI piping.</p>
<h2 id="contributions"><a class="header" href="#contributions">Contributions</a></h2>
<p>Contributions are very welcome - if you have something in mind
feel free to submit an issue, a PR, or reach out to me.</p>
<h2 id="issues"><a class="header" href="#issues">Issues</a></h2>
<p>Please address all issues to future contributors</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="installation"><a class="header" href="#installation">Installation</a></h1>
<p>This first requires installing the rust package manager <code>cargo</code>:</p>
<pre><code class="language-bash">curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
</code></pre>
<p>Afterwards we can install <code>fxtools</code> using <code>cargo</code>:</p>
<pre><code class="language-bash">cargo install fxtools
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-cat-"><a class="header" href="#-fxtools-cat-">[ <code>fxtools cat</code> ]</a></h1>
<h2 id="summary"><a class="header" href="#summary">Summary</a></h2>
<p>This command will concatenate multiple fastx files into a single stream.</p>
<h3 id="expected-input"><a class="header" href="#expected-input">Expected Input</a></h3>
<h4 id="sample_1fa"><a class="header" href="#sample_1fa"><code>sample_1.fa</code></a></h4>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
&gt;AP2S2_a
CCCT
</code></pre>
<h4 id="sample_2fa"><a class="header" href="#sample_2fa"><code>sample_2.fa</code></a></h4>
<pre><code class="language-text">&gt;NSD1_a
ACTG
&gt;NSD1_b
ACTT
&gt;NSD2_a
CCCT
</code></pre>
<h3 id="expected-output"><a class="header" href="#expected-output">Expected Output</a></h3>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
&gt;AP2S2_a
CCCT
&gt;NSD1_a
ACTG
&gt;NSD1_b
ACTT
&gt;NSD2_a
CCCT
</code></pre>
<h2 id="usage"><a class="header" href="#usage">Usage</a></h2>
<pre><code class="language-bash"># standard concatenation
fxtools cat -i &lt;fastx&gt; &lt;...&gt; &lt;fastx&gt;

# into pipeline
fxtools cat -i &lt;fastx_1&gt; &lt;fastx_2&gt; &lt;fastx_3&gt; | fxtools filter -p &quot;ACT&quot;

# print only sequences of records
fxtools cat -s -i &lt;fastx&gt; &lt;...&gt; &lt;fastx&gt;

# print only sequences of records and write as single-line
fxtools cat -Ss -i &lt;fastx&gt; &lt;...&gt; &lt;fastx&gt;

# print only headers of records
fxtools cat -s -i &lt;fastx&gt; &lt;...&gt; &lt;fastx&gt;
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-clip-"><a class="header" href="#-fxtools-clip-">[ <code>fxtools clip</code> ]</a></h1>
<h2 id="summary-1"><a class="header" href="#summary-1">Summary</a></h2>
<p>This command will clip (or truncate) the records to only recover nucleotides within
a desired range.</p>
<h3 id="expected-input-1"><a class="header" href="#expected-input-1">Expected Input</a></h3>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
&gt;AP2S2_a
CCCT
</code></pre>
<h3 id="expected-output-1"><a class="header" href="#expected-output-1">Expected Output</a></h3>
<pre><code class="language-bash"># trims 1 nucleotide from the start and 1 nucleotide from the end
fxtools clip -s 1 -e 1
</code></pre>
<pre><code class="language-text">&gt;AP2S1_a
CT
&gt;AP2S1_b
CT
&gt;AP2S2_a
CC
</code></pre>
<h2 id="usage-1"><a class="header" href="#usage-1">Usage</a></h2>
<pre><code class="language-bash"># left side clip (10 nucleotides)
fxtools clip -i &lt;fastx&gt; -s 10

# right side clip (10 nucleotides)
fxtools clip -i &lt;fastx&gt; -e 10

# left side clip (5 nucleotides) right side clip (15) nucleotides
fxtools clip -i &lt;fastx&gt; -s 5 -e 15

# clip everything outside of nucleotide range 10-20
fxtools clip -i &lt;fastx&gt; -r 10..20

# clip everything outside range 10-end (equivalent to -s 10)
fxtools clip -i &lt;fastx&gt; -r 10..

# clip everything outside range start-10
fxtools clip -i &lt;fastx&gt; -r ..10
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-count-"><a class="header" href="#-fxtools-count-">[ <code>fxtools count</code> ]</a></h1>
<h2 id="summary-2"><a class="header" href="#summary-2">Summary</a></h2>
<p>This command will just count the number of records in your input <code>fastx</code></p>
<h3 id="expected-input-2"><a class="header" href="#expected-input-2">Expected Input</a></h3>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
&gt;AP2S2_a
CCCT
</code></pre>
<h3 id="expected-output-2"><a class="header" href="#expected-output-2">Expected Output</a></h3>
<pre><code class="language-text">3
</code></pre>
<h2 id="usage-2"><a class="header" href="#usage-2">Usage</a></h2>
<pre><code class="language-bash"># standard counting
fxtools count -i &lt;fastx&gt;

# from pipeline
fxtools filter -i &lt;fastx&gt; -p &quot;ACTCGCG&quot; | fxtools count
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-extract-variable-"><a class="header" href="#-fxtools-extract-variable-">[ <code>fxtools extract-variable</code> ]</a></h1>
<h2 id="summary-3"><a class="header" href="#summary-3">Summary</a></h2>
<p>This command will extract the variable regions from an input <code>fastx</code> and write those variable regions to the output <code>fastx</code>. </p>
<h3 id="expected-input-sequences"><a class="header" href="#expected-input-sequences">Expected Input Sequences</a></h3>
<p>It was designed assuming that the sequences are all equal size and that they are prefixed and suffixed by a fairly static
nucleotide region (consider CRISPRi/a libraries with a constant adapter sequence on either side of a highly variable region).</p>
<pre><code class="language-text">[prefix][variable][suffix]
[prefix][variable][suffix]
           ...
[prefix][variable][suffix]
</code></pre>
<h3 id="expected-output-sequences"><a class="header" href="#expected-output-sequences">Expected Output Sequences</a></h3>
<p>The output sequences will extract just the positions of the input sequence that have a higher entropy than
random chance.</p>
<pre><code class="language-text">[variable]
[variable]
   ...
[variable]
</code></pre>
<h3 id="how-it-works"><a class="header" href="#how-it-works">How it Works</a></h3>
<p>This works by calculating the positional entropy across the nucleotides at each position, then applies a z-score threshold on
those entropies to determine a contiguous variable region which is then used as the bounds to write the output sequences.</p>
<h3 id="parameters"><a class="header" href="#parameters">Parameters</a></h3>
<p>Default will write to stdout, but you can provide an output file with the <code>-o</code> flag.
You can decide how many sequences to calculate the entropy on with the <code>-n</code> flag.
You can decide what z-score threshold to use for your data with the <code>-z</code> flag.</p>
<blockquote>
<p><strong>Note:</strong></p>
<p>The z-score threshold default is arbitrarily set.
If you have a smaller number of sequences try to reduce the
threshold to <code>0.5</code>, and see if that helps.</p>
</blockquote>
<h2 id="usage-3"><a class="header" href="#usage-3">Usage</a></h2>
<pre><code>fxtools extract-variable \
  -i &lt;input_fastx&gt; \
  -o &lt;output_fastx&gt; \
  -n &lt;number of sequences to use in fitting entropy [default: 5000]&gt; \
  -z &lt;zscore threshold to use [default: 1.]&gt;
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-filter-"><a class="header" href="#-fxtools-filter-">[ <code>fxtools filter</code> ]</a></h1>
<h2 id="summary-4"><a class="header" href="#summary-4">Summary</a></h2>
<p>This command will filter your input <code>fastx</code> into an output <code>fastx</code>
and retrieve all the sequences or headers that match your input pattern.</p>
<p>This pattern is a regex compatible pattern, and can also be inverted with
the <code>-v</code> flag (like <code>grep -v</code>).</p>
<h3 id="expected-input-3"><a class="header" href="#expected-input-3">Expected Input</a></h3>
<p>This will reverse complement each of the sequences and potential quality scores.</p>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
&gt;AP2S2_a
CCCT
</code></pre>
<h3 id="expected-output-3"><a class="header" href="#expected-output-3">Expected Output</a></h3>
<h4 id="filter-on-sequence"><a class="header" href="#filter-on-sequence">Filter on Sequence</a></h4>
<pre><code class="language-bash">fxtools filter -i &lt;fasta&gt; -p &quot;ACT&quot;
</code></pre>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
</code></pre>
<h4 id="filter-on-header"><a class="header" href="#filter-on-header">Filter on Header</a></h4>
<pre><code class="language-bash">fxtools filter -i &lt;fasta&gt; -p &quot;_a&quot; -H
</code></pre>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S2_a
CCCT
</code></pre>
<h4 id="inverse-filter"><a class="header" href="#inverse-filter">Inverse Filter</a></h4>
<pre><code class="language-bash">fxtools filter -i &lt;fasta&gt; -p &quot;ACT&quot; -v
</code></pre>
<pre><code class="language-text">&gt;AP2S2_a
CCCT
</code></pre>
<h2 id="usage-4"><a class="header" href="#usage-4">Usage</a></h2>
<pre><code class="language-bash"># standard filtering (on sequence)
fxtools filter -i &lt;fastx&gt; -p &lt;pattern&gt;

# filtering on header
fxtools filter -i &lt;fastx&gt; -p &lt;pattern&gt; -H

# inverse filter (removing all records that match pattern)
fxtools filter -i &lt;fastx&gt; -p &lt;pattern&gt; -v

# inverse filter (removing all records that match pattern) on header
fxtools filter -i &lt;fastx&gt; -p &lt;pattern&gt; -v -H
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-fix-"><a class="header" href="#-fxtools-fix-">[ <code>fxtools fix</code> ]</a></h1>
<h2 id="summary-5"><a class="header" href="#summary-5">Summary</a></h2>
<p>This command will replace all nucleotides not matching: <code>[ACTGNactgn]</code>
with the missing nucleotide <code>N</code>.</p>
<h3 id="parameters-1"><a class="header" href="#parameters-1">Parameters</a></h3>
<p>Default will write to stdout, but you can provide an output file with the <code>-o</code> flag.</p>
<h2 id="usage-5"><a class="header" href="#usage-5">Usage</a></h2>
<pre><code class="language-bash">fxtools fix \
  -i &lt;input_fastx&gt; \
  -o &lt;output_fastx&gt;
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-sort-"><a class="header" href="#-fxtools-sort-">[ <code>fxtools sort</code> ]</a></h1>
<h2 id="summary-6"><a class="header" href="#summary-6">Summary</a></h2>
<p>This tool can be used to sort a <code>fastx</code> file or
a pair of equivalently sized <code>fastx</code> files based on
sequence.</p>
<p>If an <code>R2</code> is provided, then the default is to sort
on its sequences.</p>
<h2 id="usage-6"><a class="header" href="#usage-6">Usage</a></h2>
<pre><code class="language-bash"># Sort a single fastq
fxtools sort -i &lt;your_file.fastq&gt;

# Sort a paired-end fastq set by R2
fxtools sort -i &lt;your_R1.fq.gz&gt; -I &lt;your_R2.fq.gz&gt;

# Sort a paired-end fastq set by R1
fxtools sort -i &lt;your_R1.fq.gz&gt; -I &lt;your_R2.fq.gz&gt; --sort-by-r1
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-reverse-"><a class="header" href="#-fxtools-reverse-">[ <code>fxtools reverse</code> ]</a></h1>
<h2 id="summary-7"><a class="header" href="#summary-7">Summary</a></h2>
<p>This command will convert your input <code>fastx</code> into an output <code>fastx</code>
with all nucleotide sequences (and associated quality scores) in
in reverse order.</p>
<p>Useful for <code>grep</code> for a sequence in <code>R2</code> or vice versa.</p>
<h3 id="expected-input-4"><a class="header" href="#expected-input-4">Expected Input</a></h3>
<p>This will reverse complement each of the sequences and potential quality scores.</p>
<pre><code class="language-text">ACTG
GCTA
AAAA
</code></pre>
<h3 id="expected-output-4"><a class="header" href="#expected-output-4">Expected Output</a></h3>
<pre><code class="language-text">CAGT
TAGC
TTTT
</code></pre>
<h2 id="usage-7"><a class="header" href="#usage-7">Usage</a></h2>
<pre><code class="language-bash">fxtools reverse -i &lt;your_fwd.fq.gz&gt;
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-sample-"><a class="header" href="#-fxtools-sample-">[ <code>fxtools sample</code> ]</a></h1>
<h2 id="summary-8"><a class="header" href="#summary-8">Summary</a></h2>
<p>This command will randomly downsample the number of records in your input <code>fastx</code>
by some frequency <code>f</code>.</p>
<h3 id="expected-input-5"><a class="header" href="#expected-input-5">Expected Input</a></h3>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
&gt;AP2S2_a
CCCT
&gt;AP2S2_b
CCCC
</code></pre>
<h3 id="expected-output-5"><a class="header" href="#expected-output-5">Expected Output</a></h3>
<p>With frequency 0.75.</p>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S2_a
CCCT
&gt;AP2S2_b
CCCC
</code></pre>
<h2 id="usage-8"><a class="header" href="#usage-8">Usage</a></h2>
<pre><code class="language-bash"># standard sampling 50% of records
fxtools sample -i &lt;fastx&gt; -f 0.5

# from pipeline subsampling 30% of records
fxtools filter -i &lt;fastx&gt; -p &quot;ACTCGCG&quot; | fxtools sample -f 0.3
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-sgrna-table-"><a class="header" href="#-fxtools-sgrna-table-">[ <code>fxtools sgrna-table</code> ]</a></h1>
<h2 id="summary-9"><a class="header" href="#summary-9">Summary</a></h2>
<p>This command will create a table mapping sgRNA names to their parent gene. </p>
<h3 id="expected-input-6"><a class="header" href="#expected-input-6">Expected Input</a></h3>
<p>This works by parsing the header of each record and currently it expects the header to be as follows:</p>
<pre><code class="language-bash"># {gene}_{auxilliary sgrna description}
</code></pre>
<h3 id="parameters-2"><a class="header" href="#parameters-2">Parameters</a></h3>
<p>The command requires an input fasta/q file and will by default write a sgrna-to-gene table to stdout.</p>
<p>You can pipe the output table to a file with the <code>-o</code> flag.</p>
<p>You can also choose to include each records sequence with the <code>-s</code> flag. </p>
<p>You can also choose to reorder the columns to whatever format you'd like with the <code>-r</code> flag
and provide a 3 character string (i.e. <code>-r hsg</code> or <code>-r ghs</code>) representing the <code>[hH]eader</code>, 
<code>[sS]sequence</code>, and <code>[gG]ene</code>.</p>
<p>By default the table's delimiter is tabs, but you can specify a separate delimiter with the <code>-d</code> flag.</p>
<h2 id="usage-9"><a class="header" href="#usage-9">Usage</a></h2>
<pre><code class="language-bash">fxtools sgrna-table \
  -i &lt;input_fastx&gt; \
  -o &lt;s2g.txt&gt; \
  -s \
  -r ghs \
  -d &lt;character delim&gt;
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-t2g-"><a class="header" href="#-fxtools-t2g-">[ <code>fxtools t2g</code> ]</a></h1>
<h2 id="summary-10"><a class="header" href="#summary-10">Summary</a></h2>
<p>This tool can be used to parse a cDNA fasta file
and build a mapping of transcripts to genes which
is used in Kallisto.</p>
<p>You can decide to include the ensembl <code>gene_id</code>
or the <code>gene_name</code> which is the common symbol
of that gene.</p>
<h2 id="usage-10"><a class="header" href="#usage-10">Usage</a></h2>
<pre><code class="language-bash"># parse the t2g and write to stdout
fxtools t2g -i &lt;your_seq.cdna.fasta.gz&gt;

# parse the t2g and write to `t2g.txt`
fxtools t2g -i &lt;your_seq.cdna.fasta.gz&gt; -o t2g.txt

# parse the t2g and write the symbols instead of the gene_id
fxtools t2g -i &lt;your_seq.cdna.fasta.gz&gt; -s

# parse the t2g and include the gene_id version in the output
fxtools t2g -i &lt;your_seq.cdna.fasta.gz&gt; -s -d
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-take-"><a class="header" href="#-fxtools-take-">[ <code>fxtools take</code> ]</a></h1>
<h2 id="summary-11"><a class="header" href="#summary-11">Summary</a></h2>
<p>This command will take a certain number of records from your input <code>fastx</code></p>
<h3 id="expected-input-7"><a class="header" href="#expected-input-7">Expected Input</a></h3>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
&gt;AP2S2_a
CCCT
</code></pre>
<h3 id="expected-output-take-2"><a class="header" href="#expected-output-take-2">Expected Output (take 2)</a></h3>
<pre><code class="language-text">&gt;AP2S1_a
ACTG
&gt;AP2S1_b
ACTT
</code></pre>
<h2 id="usage-11"><a class="header" href="#usage-11">Usage</a></h2>
<pre><code class="language-bash"># standard take (taking 3 records from the top)
fxtools take -i &lt;fastx&gt; -n 3

# standard take (skipping 2 and then taking 3 records)
fxtools take -i &lt;fastx&gt; -s 2 -n 3

# from pipeline
# filtering for a sequence pattern, and then taking the first 30 hits
fxtools filter -i &lt;fastx&gt; -p &quot;ACTCGCG&quot; | fxtools take -n 30
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-trim-"><a class="header" href="#-fxtools-trim-">[ <code>fxtools trim</code> ]</a></h1>
<h2 id="summary-12"><a class="header" href="#summary-12">Summary</a></h2>
<p>This tool can be used to select and trim sequences
that contain a specified adapter sequence.</p>
<p>This can almost be thought of as a combination of <code>grep</code> 
and <code>sed</code>, where everything before the <code>grep</code> match is
removed.</p>
<p>By default the adapter sequence is kept (the adapter
will be a common prefix for all kept reads) but it
can also be trimmed away.</p>
<h2 id="usage-12"><a class="header" href="#usage-12">Usage</a></h2>
<pre><code class="language-bash"># trim away sequences prefixing the adapter
fxtools trim -i &lt;your_seq.fq.gz&gt; -a ACTTGGA

# trim away sequences prefixing the adapter + the adapter
fxtools trim -i &lt;your_seq.fq.gz&gt; -a ACTTGGA --trim-adapter
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-unique-"><a class="header" href="#-fxtools-unique-">[ <code>fxtools unique</code> ]</a></h1>
<h2 id="summary-13"><a class="header" href="#summary-13">Summary</a></h2>
<p>This is a command that will determine all unique sequences within a fastx file and split the records into either unique or duplicates.
It will also count the number of unique/duplicate sequences/duplicate records and report those.</p>
<h3 id="parameters-3"><a class="header" href="#parameters-3">Parameters</a></h3>
<p>By default all unique reads will be pushed to stdout unless piped to a file with the <code>-o</code> flag.
Nulled reads will not be reported by default but can be written to a filepath with the <code>-n</code> flag.</p>
<h2 id="usage-13"><a class="header" href="#usage-13">Usage</a></h2>
<pre><code class="language-bash">fxtools unique \
  -i &lt;input_fastx&gt; \
  -o &lt;optional_output_file_for_unique&gt; \
  -n &lt;optional_output_file_for_null&gt;
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="-fxtools-upper-"><a class="header" href="#-fxtools-upper-">[ <code>fxtools upper</code> ]</a></h1>
<h2 id="summary-14"><a class="header" href="#summary-14">Summary</a></h2>
<p>This command will convert your input fastx into an output fastx with all nucleotides converted to their uppercase.
This will also validate to ensure there are no unexpected nucleotides found.</p>
<h3 id="parameters-4"><a class="header" href="#parameters-4">Parameters</a></h3>
<p>Default will write to stdout, but you can provide an output file with the <code>-o</code> flag.</p>
<h2 id="usage-14"><a class="header" href="#usage-14">Usage</a></h2>
<pre><code class="language-bash">fxtools upper \
  -i &lt;input_fastx&gt; \
  -o &lt;output_fastx&gt;
</code></pre>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
            </nav>

        </div>

        <script type="text/javascript">
            window.playground_copyable = true;
        </script>
        <script src="elasticlunr.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="mark.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="searcher.js" type="text/javascript" charset="utf-8"></script>
        <script src="clipboard.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="highlight.js" type="text/javascript" charset="utf-8"></script>
        <script src="book.js" type="text/javascript" charset="utf-8"></script>

        <!-- Custom JS scripts -->
        <script type="text/javascript">
        window.addEventListener('load', function() {
            window.setTimeout(window.print, 100);
        });
        </script>
    </body>
</html>
